### PR TITLE
Cross-host docker builds

### LINKED ISSUE #7741: ACKL - Docker file replacement to build and test and new functionality
New docker file to build and test the ACLK to cloud functionality

- Refresh the agent code from the repository
  - Includes the forked source code of the mosquitto library
- Build the libmosquitto static library (under netdata/mosquitto)
- Build the agent so as it statically links the libmosquitto.a library
- Configure the agent to be in a claimed state
- Configure the [agent_cloud_link] section in netdata.conf

### LINKED ISSUE #7899: Docker files for dev / testing.
#### Summary

One of the problems that we face in reviewing PRs is that there are many complex configurations for Netdata that are difficult to setup and use for testing. The current approach is to use a set of VMs that contain various OS configurations, Netdata installs and all of the other tools that we need for a full monitoring stack. The problem with this is that it is not robust, easy to maintain and does not allow repeatable testing during development.

We can overcome these issues with a custom version of the Dockerfiles that are used for packaging Netdata for different target environments. The main differences between our use-cases and packaging are:

* Must support incremental builds for dev - not always a clean install.
* Must support debugging / valgrind and other tools.
* Must support injection of partial configuration files for a predictable setup.
* Must be usable in docker-compose to create multi-node configurations.

### Expected outcome

Support the following configurations for dev/testing:

* Simple single Netdata installation (i.e. the basic one-node case).
* Master/slave configurations with multiple Netdata nodes.
* ACLK testing configuration with custom fork, VerneMQ etc.
* Monitoring stack with Prometheus / Grafana and other configs that we need to test exporting on.

### PR DESCRIPTION
**Pull request by amoss**
##### Summary
Closes #7899.

##### Component Name
Agent build system.

##### Additional Information
Incremental cross-host builds for the agent using Docker.

### CODE REVIEW

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

File: build_external/bin/clean_install.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+TagName=clean_install_"$Distro"_"$Version":latest
+Dockerfile=$Base/clean_install_"$Distro"_"$Version".Dockerfile
+
+if [ ! -e "$Dockerfile" ]; then
+    echo "Can't run installer on $Distro/$Version - $Dockerfile not found"
+    exit -1
+fi
+
+if [ "$Arch" == "Linux" ]; then
+    sudo docker build -f "$Dockerfile" "$Repo" -t "$TagName"
```
Comment (prologic) [inline build_external/bin/clean_install.sh]: `sudo` is almost never necessary with Docker in general if you have Docker setup (_which it should be by default_) in such a way as you were a member of the `docker` group and the `/var/lib/docker.sock` UNIX socket has group read/write permissions.

**Comment by prologic** [comment]
cc @Ferroin 

@amoss As discussed, I _think_ this PR has a lot of overlap with what we're doing in #7735 netdata/helper-images#31 netdata/helper-images#32 netdata/helper-images#33 netdata/helper-images#34

I _think_ we're doing very similar things but with slightly different goals. It would be awesome if we could find a way to converge on a single set of `Dockerfile(s)` at the very least so we can achieve both local dev and CI/CD.

**Comment by amoss** [comment]
I agree completely - I think we are doing very similar things with slightly different aims. If we can discuss how to merge them then I think that it's a good outcome for all of us and we can avoid some duplication / redundancy.

**Comment by prologic** [comment]
> @prologic: I agree completely - I think we are doing very similar things with slightly different aims. If we can discuss how to merge them then I think that it's a good outcome for all of us and we can avoid some duplication / redundancy.

No worries! My inclination at this point is to just let the duplication happen "for now" given our timeline and refactor it / clean it up later.

> Also, I'll check the docker permissions. I've never used it without privileges so I assumed that was the default case :)

Yeah :) No need to sun the "client" as root -- it just talks to a daemon over a UNIX socket :)

**Comment by amoss** [comment]
Of course changing group membership is awkward when you have a desktop environment, but if this trips anybody else up - here is a very hacky trick:
```
exec newgrp docker; exec newgroup $USER; exec bash -l
```

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 13
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=4828736&bid=16044683)

File: build_external/bin/clean-install.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+
+DISTRO="$1"
+VERSION="$2"
+BuildBase="$(cd "$(dirname "$0")" && cd .. && pwd)"
+
+# This is temporary - not all of the package-builder images from the helper-images repo
+# are available on Docker Hub. When everything falls under the "happy case"  below this
+# can be deleted in a future iteration. This is written in a weird way for portability,
+# can't rely on bash 4.0+ to allow case fall-through with ;&
```
Comment (Ferroin) [inline build_external/bin/clean-install.sh]: I assume you're referring to the `*_i386` variants not being on Docker hub. Those are, for the moment, a hack to allow for building of 32-bit x86 packages until `docker buildx` is more stable. Unless we specifically need handling for 32-bit distros here, we can probably safely ignore them for this.

Comment (amoss) [reply build_external/bin/clean-install.sh]: No - it was Arch Linux that is not available at all as a tag under the image.

Comment (amoss) [reply build_external/bin/clean-install.sh]: The ones in the whitelist are taken from https://hub.docker.com/r/netdata/package-builders/tags although I also checked manually what I could pull and it seemed to match this list.

Comment (Ferroin) [reply build_external/bin/clean-install.sh]: Ah, OK, that's not likely to change at any time in the future. We're not building AUR packages ourselves and it's not a high priority to take over maintainership there as the person who's maintaining that within the Arch community is honestly doing an awesome job.

**Comment by Ferroin** [comment]
I'm working on testing this locally today. I should hopefully have either an approving review or comments about errors some time later today.

File: build_external/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+infrastructure (these parts of the code are not public), or with external brokers
+(such as VerneMQ for MQTT), or with other external tools (such as TSDB to allow the agent to
+export metrics). Note: no external TSDB scenarios are available in the first iteration,
+they will be added in subsequent iterations.
+
+This differs from the packaging dockerfiles as it designed to be used for local development.
+The main difference is that these files are designed to support incremental compilation in
+the following way:
+
+1. The initial build should be performed using `bin/clean-install.sh`` to create a docker
```
Comment (mfundul) [inline build_external/README.md]: `bin/clean-install.sh`` ->  `bin/clean-install.sh`

Comment (amoss) [reply build_external/README.md]: Vscode :)

File: build_external/README.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    wth the same GUID kills the master which exits with a fatal condition.
+    
+4. The ACLK
+
+This is for internal use only as it requires access to a private repo. Clone the vernemq-docker
+repo and follow the instructions within to build an image called `vernemq`.
+
+```bash
+build_external/bin/clean-install.sh arch current  # Only needed first time
+docker-compose -f build_external/projects/aclk-testing/vernemq-compose.yml -f build__external/projects/aclk-testing/agent-compose.yml up --build
```
Comment (stelfrag) [inline build_external/README.md]: There is an extra underscore in `build__external/projects/aclk-testing/agent-compose.yml`

**Pull request merged**